### PR TITLE
Fix IndexOutOfBoundsException for MSQ window function queries with empty RAC

### PR DESCRIPTION
**Pull request by Akshat-Jain**
### Description

Currently, queries like the following give IndexOutOfBoundsException error because we attempt to create a frame writer for empty RAC.

```sql
select countryName, row_number() over () as c1
from wikipedia
where countryName in ('non-existent-country')
group by countryName, cityName, channel
```

The above query gives:

```java
 Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:374)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at org.apache.druid.query.rowsandcols.ConcatRowsAndColumns.findColumn(ConcatRowsAndColumns.java:95)
	at org.apache.druid.query.rowsandcols.semantic.DefaultColumnSelectorFactoryMaker$ColumnAccessorBasedColumnSelectorFactory.withColumnAccessor(DefaultColumnSelectorFactoryMaker.java:183)
	at org.apache.druid.query.rowsandcols.semantic.DefaultColumnSelectorFactoryMaker$ColumnAccessorBasedColumnSelectorFactory.makeDimensionSelector(DefaultColumnSelectorFactoryMaker.java:83)
	at org.apache.druid.frame.field.FieldWriters.makeStringWriter(FieldWriters.java:132)
	at org.apache.druid.frame.field.FieldWriters.create(FieldWriters.java:78)
	at org.apache.druid.frame.write.RowBasedFrameWriterFactory.makeFieldWriters(RowBasedFrameWriterFactory.java:117)
	at org.apache.druid.frame.write.RowBasedFrameWriterFactory.newFrameWriter(RowBasedFrameWriterFactory.java:76)
	at org.apache.druid.msq.querykit.WindowOperatorQueryFrameProcessor.createFrameWriterIfNeeded(WindowOperatorQueryFrameProcessor.java:407)
	at org.apache.druid.msq.querykit.WindowOperatorQueryFrameProcessor.flushAllRowsAndCols(WindowOperatorQueryFrameProcessor.java:394)
```

This PR fixes it and adds a test that fails with the above stacktrace without this PR's change.

<hr>

This PR has:

- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java [hunk 1]
```diff
@@ -203,7 +203,9 @@ public ReturnOrAwait<Object> runIncrementally(IntSet readableInputs)
         final Frame frame = inputChannel.read();
         convertRowFrameToRowsAndColumns(frame);
       } else if (inputChannel.isFinished()) {
-        runAllOpsOnMultipleRac(frameRowsAndCols);
+        if (!frameRowsAndCols.isEmpty()) {
+          runAllOpsOnMultipleRac(frameRowsAndCols);
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java:207]: Shouldn't the check be inside runAllOpsMulitpleRAC ?

Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java]: Have made the change.

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java [hunk 2]
```diff
@@ -326,6 +326,9 @@ public Closeable goOrContinue(Closeable continuationObject, Receiver receiver)
    */
   private void runAllOpsOnMultipleRac(ArrayList<RowsAndColumns> listOfRacs)
   {
+    if (listOfRacs.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (Akshat-Jain) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java]: Have moved the change to `ConcatRowsAndColumns`, and added a test in `ConcatRowsAndColumnsTest` which fails with `IndexOutOfBoundsException` without the code changes.

Hope that works.

File: processing/src/main/java/org/apache/druid/query/rowsandcols/ConcatRowsAndColumns.java [hunk 3]
```diff
@@ -92,6 +92,9 @@ public Column findColumn(String name)
     if (columnCache.containsKey(name)) {
       return columnCache.get(name);
     } else {
+      if (racBuffer.isEmpty()) {
```
Comment (kgyrtkirk) [inline processing/src/main/java/org/apache/druid/query/rowsandcols/ConcatRowsAndColumns.java:95]: while we are at this; can you also add a `Preconditions` check to the constructor?

please also check&test other possibly affected methods as well ; I guess `getColumnNames()` would throw the same error

Comment (Akshat-Jain) [reply processing/src/main/java/org/apache/druid/query/rowsandcols/ConcatRowsAndColumns.java:101]: @kgyrtkirk Done

**Review by kgyrtkirk — APPROVED** [review]
thank you for the updates!

**Pull request merged**